#947 filter_warning option to replace default "No needs passed the filters" text

kreuzberger opened the pull request:
Add option `:filter_warning:` to directives (e.g. needtable) to show no text or given text instead of the default text.

danwos commented on sphinx_needs/directives/needextract.py:
```diff
@@ -120,7 +120,11 @@ def process_needextract(
-            content.append(no_needs_found_paragraph())
+            content.append(
+                no_needs_found_paragraph(
+                    current_needextract["filter_warning"] if "filter_warning" in current_needextract else None
```
I guess this can be shorten to:
```
current_needextrac.get(["filter_warning", None)
```

danwos commented on sphinx_needs/directives/needfilter.py:
```diff
@@ -228,7 +228,11 @@ def process_needfilters(
 
         if len(content) == 0:
-            nothing_found = "No needs passed the filters"
+            nothing_found = (
```
This we could write like 
```
nothing_found = current_needfilter.get('filter_warning', 'No needs passed the filters')
```

danwos commented on tests/doc_test/filter_doc/filter_no_needs.rst:
```diff
@@ -0,0 +1,26 @@
+
+.. needtable::
+   :filter: ("7" in tags)
+   :filter_warning: 
```
Would add here a positive filter result as well, to be sure the text is not printed if needs got found.

danwos reviewed (changes requested):
Thanks for the PR. :+1: 
Technical realization looks good.

I think the dictionary operations can be shortened. I left 1-2 comments, but it affects all changes for the different directives.

kreuzberger commented:
Hi! Just your suggestions will be integrated cause they provide for sure a more pythonic solution.
Just 2 additional questions:
- the lint gives error due to some apply method i do not know nothing about / unchanged code by my pr. How to solve this?
- Technically some directives (e.g. needtables) seem to give this warning message or the filter_warning text, but some directives (e.g. needlist) seem to remove the whole node / content even if the code for the warning message is active? Is this right? intended?

danwos commented:
Regarding the linter problem, this may happen if mypy or other linter-libs got an update and introduced new checks.

However, you can deactivate the specific test in `pyproject.toml (untested!)`:
```toml
[[tool.mypy.overrides]]
module = [
  "sphinx_needs.directives.needextract",
]
disable_error_code = "no-untyped-call"
```

Regarding `needlists`, I guess this is a bug or not updated behavior.
For me, writing the "no needs found"-text for all directives should be the common rule.
So if you like, you can fix it in this PR  ;)

Thanks for the invested time!! :pray:

kreuzberger commented:
So, made some fixes:

- add you hints regarding code style
- fix the mypy issue regarding your comments
- fix different handling of filtered data implementations in most directives

Changed behaviour:

- all directives use the common method ``directives.utils.no_needs_found_paragraph``
If no needs are found due to filter, the default or filter warning message is shown.
This includes that if the warning message is shown the "empty" content is not shown!! This affects:
- Tables: no more empty tables shown
- Lists: show filter message in the above cases (was not shown before)
- needgantt: remove empty figure nodes.
- needsequence: remove figure nodes if 0 == messages  and just one participant.

Added more tests, including a good case there needs are found and filtered message should not be shown.
[filter_no_needs.pdf](https://github.com/useblocks/sphinx-needs/files/14008828/filter_no_needs.pdf)

kreuzberger commented:
The checks failed due to upload error to codecov...

kreuzberger commented:
Just one more questions:
If i run the tests locally, (e.g. `nox -s "tests-3.10(sphinx='7.0')" `) i always get an error in this test:
`tests/test_sn_collapse_button.py FF                                                                                                                                                  [ 90%]`

The error message shown is
```
Timing measurement results (JSON) stored under /tmp/sn_test_build_data/pgkTg3kpnD/variant_doc/_build/html/debug_measurement.json
Timing measurement report (HTML) stored under /tmp/sn_test_build_data/pgkTg3kpnD/variant_doc/_build/html/debug_measurement.html
No version of Cypress is installed in: ~/.cache/Cypress/13.6.3/Cypress

Please reinstall Cypress by running: cypress install

----------

Cypress executable not found at: ~/.cache/Cypress/13.6.3/Cypress/Cypress
```
Is there a specific problem with the pyvenv/nox or what could get wrong?

danwos commented:
Cypress is our frontend JavaScript test framework, to test browser based features like the need-collapse button.
It must be installed by hand.

I guess there is a flag for all frontend tests, so it could be deactivate with the right pytest command.

danwos reviewed (approved):
Looks good. Thanks for the update and the good tests.

I will work a little bit on the docs and merge it then (hopefully today)

kreuzberger commented:
With the latest commit i think everything is now successfully finished. If you have no more comments/issues with the PR, i think it is ready to get merged :smile:

danwos reviewed (approved):
Looks good, thanks a lot for the PR :+1:
